src/epgsql_wire.erl

```diff
@@ -310,6 +316,39 @@ encode_standby_status_update(ReceivedLSN, FlushedLSN, AppliedLSN) ->
     Timestamp = ((MegaSecs * 1000000 + Secs) * 1000000 + MicroSecs) - 946684800*1000000,
     <<$r:8, ReceivedLSN:?int64, FlushedLSN:?int64, AppliedLSN:?int64, Timestamp:?int64, 0:8>>.
 
+%% @doc encode binary copy data file header
+%%
+%% See [https://www.postgresql.org/docs/current/sql-copy.html#id-1.9.3.55.9.4.5]
+encode_copy_header() ->
+    <<
+      "PGCOPY\n", 8#377, "\r\n", 0,             % "signature"
+      0:?int32,                                 % flags
+      0:?int32                                  % length of the extensions area
+    >>.
+
+%% @doc encode binary copy data file row / tuple
+%%
+%% See [https://www.postgresql.org/docs/current/sql-copy.html#id-1.9.3.55.9.4.6]
+encode_copy_row(ValuesTuple, Types, Codec) when is_tuple(ValuesTuple) ->
+    encode_copy_row(tuple_to_list(ValuesTuple), Types, Codec);
+encode_copy_row(Values, Types, Codec) ->
+    NumCols = length(Types),
+    [<<NumCols:?int16>>
+    | [
+       case epgsql_binary:is_null(Value, Codec) of
+           true ->
+               <<-1:?int32>>;
+           false ->
+               epgsql_binary:encode(Type, Value, Codec)
+       end || {Type, Value} <- lists:zip(Types, Values) % TODO: parallel iteration ninstead
+      ]].
```

Comment:
nittest of nits: consider optimizing the `lists:zip` away by recursively calling a function and matching on the heads of Values and Types to produce these.

Reply:
Yeah.. Also, even better optimization could be achieved by implementation of a function similar to `epgsql_wire:build_decoder`. Because right now we do a lots of internal "codec" table lookups on each data "cell" and it can be cached, because each data row has the same structure. This optimization could be reused by `epgsql:execute_batch` as well.
I'm thinking about making a separate PR with all those optimizations.

Reply:
Just accidentally discovered there exists `lists:zipwith/3`. So, Can be rewritten as

```suggestion
| lists:zipwith(
    fun(Type, Value) ->
        case epgsql_binary:is_null(Value, Codec) of
           true ->
               <<-1:?int32>>;
           false ->
               epgsql_binary:encode(Type, Value, Codec)
       end
    end, Types, Values)].
```